bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/link/ItemChannelLinkResource.java

```diff
@@ -236,9 +236,10 @@ public Response link(@PathParam("itemName") @Parameter(description = "itemName")
             }
             if (!(profileType.getSupportedItemTypes().isEmpty()
                     || profileType.getSupportedItemTypes().contains(itemType))
-                    || !(((TriggerProfileType) profileType).getSupportedChannelTypeUIDs().isEmpty()
-                            || ((TriggerProfileType) profileType).getSupportedChannelTypeUIDs()
-                                    .contains(channel.getChannelTypeUID()))) {
+                    || !(profileType.getSupportedChannelTypeUIDs().isEmpty()
+                            || profileType.getSupportedChannelTypeUIDs().contains(channel.getChannelTypeUID()))
+                    || !(profileType.getSupportedItemTypesOfChannel().isEmpty()
+                            || profileType.getSupportedItemTypesOfChannel().contains(itemType))) {
```

Comment:
That looks really horrible. It would probably be better to 

```
   Collection<String> supportedItemTypes = profileType.getSupportedItemTypes();
   ...
   if ((!supportedItemType.isEmpty() && !supportedItemTypes.contains(itemType)) || ...) {

```

Why do you check `profileType.getSupportedItemTypesOfChannel`? A profile can change the item-type:

- The timestamp update profile works with all channel-types (e.g. a humidity channel with supported item-type `Number:Dimensionless`) but can only be linked to `DateTime` items. So a check would fail (in fact, currently the profile does not provide information about accepted channel types, so this will not cause an issue, but in principle it could).
- A profile could map numeric values to strings, so the channel would support `Number`, but the profile `String`. In this case you could never link them.

Reply:
I agree that it looks horrible, but it was jut to "fit in" with the already existing code.

As to why I do the check, well it's because I did a global search for the interface methods and made sure both methods are called in all the places where one was called following an `instanceof` call.

Reply:
I agree, the style was not good before either and if you want to keep it: ok. But the `getSupportedItemTypesOfChannel` is excluding valid combinations.

Reply:
I'm definitely in favor of changing the code style here, but as always when I'm submitting a pull request, I try to limit my changes to the code logic, adapting to the local style I encounter.
I'll try to come up with a more sensible way of writing this up.